pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,248 @@
+}
+
+func (tb *Batch) add(in InsertTracesReq) {
+	_, addReqSpan := tracer.Default().Start(tb.ctx, "add-trace-req-to-batch",
+		trace.WithLinks(trace.Link{SpanContext: in.spanCtx}))
+	defer addReqSpan.End()
+	inSpans := in.payload.SpanCount()
+	if inSpans == 0 {
+		return
+	}
+	in.payload.ResourceSpans().MoveAndAppendTo(tb.traces.ResourceSpans())
+	tb.spanCount += inSpans
+	tb.reqStatus = append(tb.reqStatus, in.err)
+}
+
+func (tb *Batch) isFull() bool {
+	return tb.spanCount >= tb.maxSize
+}
+
+func (tb *Batch) isEmpty() bool {
+	return tb.spanCount == 0
+}
+
+// Batcher batches trace requests and sends batches to batch writer
+// This is done to achieve better ingest performance especially b/c
+// Jaeger collector sends traces one by one
+type Batcher struct {
+	in              chan InsertTracesReq
+	stop            chan struct{}
+	batchWriter     *batchWriter
+	once            sync.Once
+	bufferedBatches chan Batch
+	wg              sync.WaitGroup
+	config          BatcherConfig
+}
+
+type BatcherConfig struct {
+	Batchers     int
+	Writers      int
+	MaxBatchSize int
+	BatchTimeout time.Duration
+}
+
+type WriteTraces func(context.Context, ptrace.Traces) error
+
+func NewBatcher(config BatcherConfig, writer Writer) *Batcher {
+	if config.Batchers == 0 {
+		config.Batchers = defaultBatchWorkers
+	}
+	if config.MaxBatchSize == 0 {
+		config.MaxBatchSize = DefaultBatchSize
+	}
+	if config.BatchTimeout == 0 {
+		config.BatchTimeout = DefaultBatchTimeout
+	}
+	if config.Batchers < 1 || config.Writers < 1 {
+		panic("number of batchers and writeres must be greater then zero")
+	}
+	bufferedBatches := make(chan Batch, defaultMaxBufferedBatches)
+	return &Batcher{
+		batchWriter:     newBatchWriter(config.Writers, writer, bufferedBatches),
+		in:              make(chan InsertTracesReq, defaultReqBufferSize),
+		stop:            make(chan struct{}, 1),
+		bufferedBatches: bufferedBatches,
+		config:          config,
+	}
+}
+
+func (b *Batcher) Run() {
+	for i := 0; i < b.config.Batchers; i++ {
+		b.wg.Add(1)
+		go func() {
+			defer b.wg.Done()
+			b.batch()
+		}()
+	}
+	b.batchWriter.run()
+}
+
+func (b *Batcher) batch() {
+	_, batcherSpan := tracer.Default().Start(context.Background(), "create-batches")
+	ticker := time.NewTicker(b.config.BatchTimeout)
+	batch := NewBatch(b.config.MaxBatchSize)
+	batcherSpan.AddEvent("New Batch")
+	flushBatch := func(batch *Batch) *Batch {
+		batchCp := *batch
+		b.bufferedBatches <- batchCp
+		batcherSpan.AddEvent("Batch sent to buffer")
+		batcherSpan.AddEvent("New Batch")
+		return NewBatch(b.config.MaxBatchSize)
+	}
+	for {
+		select {
+		case <-b.stop:
+			ticker.Stop()
+			if !batch.isEmpty() {
+				flushBatch(batch)
+			}
+			return
+		case item := <-b.in:
+			if batch.isFull() {
+				batch = flushBatch(batch)
+				ticker.Reset(b.config.BatchTimeout)
+			}
+			batch.add(item)
+		case <-ticker.C:
+			batcherSpan.AddEvent("Batch timeout reached")
+			if !batch.isEmpty() {
+				batch = flushBatch(batch)
+			}
```

Comment:
This won't work as expected.
It will tick even if the very previous moment we did the `item := <-b.in` case and then flushed the batch. This is because the ticker is just created once, so it does not care about most recent event.

We should use `<-time.After(config.BatchTimeout)` in this case.